ibm_ace/datadog_checks/ibm_ace/subscription.py

```diff
@@ -0,0 +1,184 @@
+        )
+
+        self._last_execution_time = None
+        self._sub = None
+
+    @abstractmethod
+    def collect(self):
+        pass
+
+    @abstractmethod
+    def get_message_id(self, message):
+        pass
+
+    def get_latest_messages(self):
+        if self._last_execution_time is None:
+            self._last_execution_time = get_timestamp()
+
+        message_cache = {}
+        # Use as an ordered set
+        unknown_errors = {}
+
+        while True:
+            try:
+                payload = self.sub.get(None, pymqi.md(), self._get_options())
+            except pymqi.MQMIError as e:
+                if not (e.comp == pymqi.CMQC.MQCC_FAILED and e.reason == pymqi.CMQC.MQRC_NO_MSG_AVAILABLE):
+                    unknown_errors[str(e)] = traceback.format_exc()
+                    time.sleep(1)
+            else:
+                message = self.parse_message(payload)
+                message_cache[self.get_message_id(message)] = message
+
+            if self._get_elapsed_time() >= SNAPSHOT_UPDATE_INTERVAL:
+                break
+
+        tags = [f'subscription:{self.TYPE}', *self.tags]
+        if unknown_errors:
+            self._submit_health_status(ServiceCheck.CRITICAL, tags)
+            for error in unknown_errors.values():
+                self.check.log.error('Subscription error for topic string: %s\n%s', self.TOPIC_STRING, error)
+        elif not message_cache:
+            self._submit_health_status(ServiceCheck.WARNING, tags)
+            self.check.log.warning('Subscription found nothing for topic string: %s', self.TOPIC_STRING)
+        else:
+            self._submit_health_status(ServiceCheck.OK, tags)
+
+        messages = list(message_cache.values())
+        self.check.gauge('messages.current', len(messages), tags=tags)
+        self.check.log.trace('Messages found for subscription type `%s`: %s', self.TYPE, messages)
+
+        self._last_execution_time = get_timestamp()
+        return messages
+
+    def parse_message(self, payload):
+        return json.loads(payload.decode('utf-8'))
+
+    @property
+    def sub(self):
+        if self._sub is None:
+            sub = pymqi.Subscription(self.check.queue_manager)
+            # https://dsuch.github.io/pymqi/examples.html#how-to-subscribe-to-topics-and-avoid-mqrc-sub-already-exists-at-the-same-time
+            self.check.log.info('Subscribing to topic string: %s', self.TOPIC_STRING)
+            sub.sub(
+                sub_name=self.name,
+                topic_string=self.TOPIC_STRING,
+                sub_opts=(
+                    pymqi.CMQC.MQSO_CREATE + pymqi.CMQC.MQSO_RESUME + pymqi.CMQC.MQSO_DURABLE + pymqi.CMQC.MQSO_MANAGED
+                ),
+            )
+            self._sub = sub
+
+        return self._sub
+
+    def disconnect(self):
+        if self._sub is not None:
+            self._sub.close(sub_close_options=pymqi.CMQC.MQCO_KEEP_SUB, close_sub_queue=True)
```

Comment:
I'm curious why you have `MQCO_KEEP_SUB`? According to https://www.ibm.com/docs/en/ibm-mq/7.5?topic=calls-mqclose-close-object `MQCO_KEEP_SUB` still keeps the subscription. Would this be a memory leak concern?

Reply:
This should be fine, I think this refers to MQ itself